File: Advent of Code 2021/lib/day1.ex
defmodule Day1 do
end

{:ok, file} = File.read("inputs/day1.in")

depths =
  String.split(file)
  |> Enum.map(&String.to_integer/1)

depths
|> Enum.reduce({0, Enum.at(depths, 0)}, fn x, acc ->
  # acc = {n_increases, prev_depth}
  if elem(acc, 1) < x do
    {elem(acc, 0) + 1, x}
  else
    {elem(acc, 0), x}
  end
end)
|> elem(0)
|> IO.inspect(label: 'Part 1')

depths
|> Enum.reduce({0, Enum.at(depths, 0), Enum.at(depths, 0), Enum.at(depths, 0)}, fn x, acc ->
  # acc = {n_increases, prev_depth, prev_depth, prev_depth}
  if elem(acc, 1) + elem(acc, 2) + elem(acc, 3) < elem(acc, 2) + elem(acc, 3) + x do
    {elem(acc, 0) + 1, elem(acc, 2), elem(acc, 3), x}
  else
    {elem(acc, 0), elem(acc, 2), elem(acc, 3), x}
  end
end)
|> elem(0)
|> IO.inspect(label: 'Part 2')

File: Advent of Code 2020/day1.py
nums = [int(s) for s in open("inputs/day1.in")]


def f(nums, t):
    for i in nums:
        if t - i in nums:
            return i
    return -1


# part 1
print("part 1")
i = f(nums, 2020)
print(i, 2020 - i, i * (2020 - i))

# part 2
print()
print("part 2")
for x in nums:
    y = f(nums, 2020 - x)
    if y != -1:
        print(x, y, 2020 - x - y, x * y * (2020 - x - y))
        break
